libs/core/execution_base/include/hpx/execution_base/traits/coroutine_traits.hpp

```diff
@@ -0,0 +1,223 @@
+//  Copyright (c) 2022 Shreyas Atre
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
```

Comment:
I think all of this file should be wrapped into a 
```
#include <hpx/config.hpp>

#if defined(HPX_HAVE_CXX20_COROUTINES)
...
#endif
```

Reply:
I had experimented with it but it didn't work. I have coroutines support yet, `HPX_HAVE_CXX20_COROUTINES` is not defined.

Reply:
```
-- Performing Test HPX_WITH_CXX20_COROUTINES - Failed
-- Performing Test HPX_WITH_CXX20_EXPERIMENTAL_SIMD - Failed
-- Performing Test HPX_WITH_CXX20_LAMBDA_CAPTURE - Success
```

Reply:
I have loaded the module llvm/12.0.1

Reply:
```
[satacker@medusa01 build]$ git diff
diff --git a/cmake/tests/cxx20_coroutines.cpp b/cmake/tests/cxx20_coroutines.cpp
index bba9190fac..460dee3609 100644
--- a/cmake/tests/cxx20_coroutines.cpp
+++ b/cmake/tests/cxx20_coroutines.cpp
@@ -5,7 +5,13 @@
 //  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
 // This tests whether C++20 coroutines are supported
+#if __has_include(<coroutine>)
 #include <coroutine>
+namespace coro = std;
+#else
+#include <experimental/coroutine>
+namespace coro = std::experimental;
+#endif
 
 struct resumable
 {
@@ -19,12 +25,12 @@ struct resumable
             return {};
         }
 
-        std::suspend_never initial_suspend() const noexcept
+        coro::suspend_never initial_suspend() const noexcept
         {
             return {};
         }
 
-        std::suspend_never final_suspend() const noexcept
+        coro::suspend_never final_suspend() const noexcept
         {
             return {};
         }
@@ -41,7 +47,7 @@ struct resumable
     {
         return 42;
     }
-    void await_suspend(std::coroutine_handle<promise_type>) const noexcept {}
+    void await_suspend(coro::coroutine_handle<promise_type>) const noexcept {}
 };
 
 resumable test()
```
Made it work

Reply:
You might want to make this safer by saying:
```
#if defined(__has_include)
#if __has_include(<coroutine>)
...
#endif
#endif
```
as some older compilers don't support `__has_include()`